### ISSUE TITLE
Unicode crash in Windows console

### ISSUE THREAD

**Issue by klonuo** [ipython/ipython#9619]
Most common code page for Windows console is 437, IMHO - at least I get that by installing Windows 10 with default options.

Have a look at this session:

```
C:\temp>chcp
Active code page: 437

C:\temp>python
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:54:25) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> г
  File "<stdin>", line 1
    ?
    ^
SyntaxError: invalid syntax
>>>

C:\temp>ipython
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:54:25) [MSC v.1900 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 5.0.0.dev -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: г
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-743a0d7767e0> in <module>()
----> 1 ---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
c:\python35\lib\site-packages\IPython\core\interactiveshell.py in run_code(self, code_obj, result)
   2928                 #rprint('Running code', repr(code_obj)) # dbg
-> 2929                 exec(code_obj, self.user_global_ns, self.user_ns)
   2930             finally:

<ipython-input-1-743a0d7767e0> in <module>()
----> 1 Traceback (most recent call last):
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 2929, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-1-743a0d7767e0>", line 1, in <module>
    \u0433
NameError: name '\u0433' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 2875, in run_ast_nodes
    if self.run_code(code, result):
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 2946, in run_code
    self.showtraceback()
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 1875, in showtraceback
    self._showtraceback(etype, value, stb)
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 1893, in _showtraceback
    print(self.InteractiveTB.stb2text(stb))
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 40, in write
    self.__convertor.write(text)
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 141, in write
    self.write_and_convert(text)
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 166, in write_and_convert
    self.write_plain_text(text, cursor, start)
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 174, in write_plain_text
    self.wrapped.write(text[start:end])
  File "c:\python35\lib\encodings\cp437.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_map)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u0433' in position 0: character maps to <undefined>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python35\Scripts\ipython-script.py", line 9, in <module>
    load_entry_point('ipython==5.0.0.dev0', 'console_scripts', 'ipython')()
  File "c:\python35\lib\site-packages\IPython\__init__.py", line 119, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "c:\python35\lib\site-packages\traitlets\config\application.py", line 596, in launch_instance
    app.start()
  File "c:\python35\lib\site-packages\IPython\terminal\ipapp.py", line 348, in start
    self.shell.mainloop()
  File "c:\python35\lib\site-packages\IPython\terminal\ptshell.py", line 457, in mainloop
    self.interact()
  File "c:\python35\lib\site-packages\IPython\terminal\ptshell.py", line 448, in interact
    self.run_cell(code, store_history=True)
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 2765, in run_cell
    interactivity=interactivity, compiler=compiler, result=result)
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 2894, in run_ast_nodes
    self.showtraceback()
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 1875, in showtraceback
    self._showtraceback(etype, value, stb)
  File "c:\python35\lib\site-packages\IPython\core\interactiveshell.py", line 1893, in _showtraceback
    print(self.InteractiveTB.stb2text(stb))
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 40, in write
    self.__convertor.write(text)
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 141, in write
    self.write_and_convert(text)
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 166, in write_and_convert
    self.write_plain_text(text, cursor, start)
  File "c:\python35\lib\site-packages\colorama\ansitowin32.py", line 174, in write_plain_text
    self.wrapped.write(text[start:end])
  File "c:\python35\lib\encodings\cp437.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_map)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u0433' in position 0: character maps to <undefined>

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@scipy.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True
```

So probably IPython should not crash on this. If I set `chcp 65001` all is fine:

```
C:\temp>chcp 65001
Active code page: 65001

C:\temp>ipython
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:54:25) [MSC v.1900 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 5.0.0.dev -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: г
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-743a0d7767e0> in <module>()
----> 1 г

NameError: name 'г' is not defined
```

Interestingly on IPython 4.2.0 if I type same letter I get help page (as if I typed <kbd>?</kbd> - the same way Python interpreter sees this)

**Comment by klonuo** [comment]
I was thinking would it be brutal to change stdout code page to 65001 when IPython starts, then revert it back to initial on exit?

**Comment by takluyver** [comment]
I think it's only now possible to actually enter characters beyond your code page, so it's something we've never had to handle before. We should catch this. I don't think changing the code page is the right thing to do - that's liable to confuse things more.

We talked about integrating @Drekin's win_unicode_console, which I think would properly fix this, i.e. allowing arbitrary unicode to be displayed.

**Comment by klonuo** [comment]
That is more sensible, great. 
So it's planned for 5.1 #8669

**Comment by klonuo** [comment]
Following WUC instructions, it seems it works just fine out of the box:

```diff
--- C:\Python35\Lib\site-packages\IPython\terminal\ptshell.py
+++ (clipboard)
@@ -24,6 +24,8 @@
 from prompt_toolkit.keys import Keys
 from prompt_toolkit.layout.processors import ConditionalProcessor, HighlightMatchingBracketProcessor
 from prompt_toolkit.styles import PygmentsStyle, DynamicStyle
+
+import win_unicode_console
 
 from pygments.styles import get_style_by_name, get_all_styles
 from pygments.token import Token
@@ -388,6 +390,7 @@
             return
 
         import colorama
+        win_unicode_console.enable()
         colorama.init()
 
         # For some reason we make these wrappers around stdout/stderr.
```

**Comment by takluyver** [comment]
Cool - do you want to make a PR with that? Maybe we can get it in for 5.0.

One change - put the import inside the function. There's no need to depend on WUC or colorama on other platforms, and that's the real reason there's a Windows check there.

**Comment by klonuo** [comment]
Ok, should I import with try/except?
I ask because colorama is imported clean, perhaps it's added as dependency in PTK...

**Comment by takluyver** [comment]
It's imported clean, but look a few lines above - if we're not on Windows, we return from this method before importing it.

**Comment by klonuo** [comment]
But who guaranties that there is colorama or WUC? Wont it raise exception otherwise?

**Comment by takluyver** [comment]
colorama is listed as a Windows dependency here: https://github.com/ipython/ipython/blob/5b668f2498c640364d1d9b74599f39a5517bd821/setup.py#L212

You'll want to add WUC to that list, so we require it on Windows.

**Comment by klonuo** [comment]
Yep :)

**Comment by Carreau** [comment]
closed by #9621
